core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4080,6 +4080,46 @@ public static List reverse(List list) {
     return list;
   }
 
+  /** SQL {@code ARRAY_TO_STRING(array, delimiter)} function. */
+  public static String arrayToString(List list, String delimiter) {
+    return arrayToString(list, delimiter, null);
+  }
+
+  /** SQL {@code ARRAY_TO_STRING(array, delimiter, nullText)} function. */
+  public static String arrayToString(List list, String delimiter, @Nullable String nullText) {
+    StringBuilder sb = new StringBuilder();
+    int size = list.size();
+    for (int i = 0; i < size; i++) {
+      boolean shouldAddDelimiter = i != 0 && sb.length() > 0;
+      Object item = list.get(i);
+      String str;
+      if (item == null) {
+        str = null;
+      } else if (item instanceof String) {
+        str = (String) item;
+      } else if (item instanceof ByteString) {
+        str = item.toString();
+      } else {
+        throw new IllegalStateException(
+            "Only String or ByteString are supported but get " + item.getClass().getName());
+      }
+      if (str == null) {
```

Comment:
Perhaps we can simplify this code and avoid duplication by doing something like
```
if (str == null) {
  if (nullText == null) {
    continue;
  }
  else {
    str = nullText;
  }
}

if (addDelimiter) {
  sb.append(delimiter);
}
sb.append(str);
```
Wdyt?